File: Sources/PalmierPro/Audio/AudioMeter.swift
import Accelerate
import AVFoundation
import Foundation
import Observation

struct AudioMeterAnalysis: Sendable, Equatable {
    struct Channel: Sendable, Equatable {
        let peak, rms: Float
    }
    let left, right: Channel
    static let silence = AudioMeterAnalysis(
        left: Channel(peak: 0, rms: 0),
        right: Channel(peak: 0, rms: 0)
    )
}

struct AudioMeterChannelDisplay: Sendable, Equatable {
    let levelDb, peakDb: Float
    let clipped: Bool
}

struct StereoAudioMeterDisplay: Sendable, Equatable {
    let left, right: AudioMeterChannelDisplay
}

struct AudioMeterChannelState: Sendable {
    nonisolated static let floorDb: Float = -60
    nonisolated static let ceilingDb: Float = 0
    nonisolated static let levelDecayDbPerSecond: Float = 24
    nonisolated static let peakDecayDbPerSecond: Float = 18
    nonisolated static let peakHoldSeconds: TimeInterval = 1.5

    private var levelDb = floorDb
    private var levelTime: TimeInterval = 0
    private var peakDb = floorDb
    private var peakHoldUntil: TimeInterval = 0
    private(set) var clipped = false

    mutating func ingest(rms: Float, peak: Float, at time: TimeInterval) {
        let current = display(at: time)
        levelDb = max(Self.decibels(rms), current.levelDb)
        levelTime = time

        let incomingPeak = Self.decibels(peak)
        if incomingPeak >= current.peakDb {
            peakDb = incomingPeak
            peakHoldUntil = time + Self.peakHoldSeconds
        } else if time > peakHoldUntil {
            peakDb = current.peakDb
            peakHoldUntil = time
        }
        clipped = clipped || peak >= 1
    }

    func display(at time: TimeInterval) -> AudioMeterChannelDisplay {
        let levelElapsed = Float(max(0, time - levelTime))
        let peakElapsed = Float(max(0, time - peakHoldUntil))
        return AudioMeterChannelDisplay(
            levelDb: max(Self.floorDb, levelDb - levelElapsed * Self.levelDecayDbPerSecond),
            peakDb: max(Self.floorDb, peakDb - peakElapsed * Self.peakDecayDbPerSecond),
            clipped: clipped
        )
    }

    mutating func resetClipping() { clipped = false }
    nonisolated static func decibels(_ amplitude: Float) -> Float {
        amplitude > 0 ? max(floorDb, 20 * log10(amplitude)) : floorDb
    }
}

@Observable
@MainActor
final class AudioMeterHub {
    private var left = AudioMeterChannelState()
    private var right = AudioMeterChannelState()

    func ingest(_ analysis: AudioMeterAnalysis, at time: TimeInterval = ProcessInfo.processInfo.systemUptime) {
        left.ingest(rms: analysis.left.rms, peak: analysis.left.peak, at: time)
        right.ingest(rms: analysis.right.rms, peak: analysis.right.peak, at: time)
    }
    func display(at time: TimeInterval = ProcessInfo.processInfo.systemUptime) -> StereoAudioMeterDisplay {
        StereoAudioMeterDisplay(left: left.display(at: time), right: right.display(at: time))
    }
    func resetClipping() {
        left.resetClipping()
        right.resetClipping()
    }
    func reset() {
        left = AudioMeterChannelState()
        right = AudioMeterChannelState()
    }
}

enum AudioLevelAnalyzer {
    nonisolated static func analyze(left: [Float], right: [Float], range: Range<Int>) -> AudioMeterAnalysis {
        let upper = min(range.upperBound, min(left.count, right.count))
        let lower = max(0, min(range.lowerBound, upper))
        guard lower < upper else { return .silence }
        let leftResult = left.withUnsafeBufferPointer { metrics($0.baseAddress! + lower, count: upper - lower) }
        let rightResult = right.withUnsafeBufferPointer { metrics($0.baseAddress! + lower, count: upper - lower) }
        return AudioMeterAnalysis(left: leftResult, right: rightResult)
    }

    nonisolated static func analyze(_ buffer: AVAudioPCMBuffer) -> AudioMeterAnalysis {
        guard buffer.format.commonFormat == .pcmFormatFloat32,
              !buffer.format.isInterleaved,
              buffer.format.channelCount > 0,
              let channels = buffer.floatChannelData,
              buffer.frameLength > 0
        else { return .silence }
        let count = Int(buffer.frameLength)
        let right = min(1, Int(buffer.format.channelCount) - 1)
        return AudioMeterAnalysis(
            left: metrics(channels[0], count: count),
            right: metrics(channels[right], count: count)
        )
    }

    nonisolated private static func metrics(_ samples: UnsafePointer<Float>, count: Int) -> AudioMeterAnalysis.Channel {
        var peak: Float = 0
        var rms: Float = 0
        vDSP_maxmgv(samples, 1, &peak, vDSP_Length(count))
        vDSP_rmsqv(samples, 1, &rms, vDSP_Length(count))
        return AudioMeterAnalysis.Channel(peak: peak, rms: rms)
    }
}

File: platforms/windows/src/PalmierPro.Core/Audio/AudioMeter.cs
namespace PalmierPro.Core.Audio;

/// One block's peak per channel handed to <see cref="AudioMeterHub"/> — mirrors the shape of the
/// Mac's `AudioMeterAnalysis` (Sources/PalmierPro/Audio/AudioMeter.swift). The Windows native tap
/// (PE_TimelineGetAudioLevels) also measures RMS per channel, but the Mac's ballistics never
/// consume it, so it stops at PalmierPro.Services.Engine.AudioLevels and isn't threaded through here.
public readonly record struct AudioMeterAnalysis(float LeftPeak, float RightPeak)
{
    public static readonly AudioMeterAnalysis Silence = new(0, 0);
}

public readonly record struct AudioMeterChannelDisplay(float LevelDb, float PeakDb, bool Clipped);

public readonly record struct StereoAudioMeterDisplay(AudioMeterChannelDisplay Left, AudioMeterChannelDisplay Right);

/// Verbatim port of `Audio/AudioMeter.swift`'s `AudioMeterChannelState` — same floor/ceiling/decay
/// constants and the same ballistics: level jumps up to an incoming peak then decays linearly
/// (dB/s), a separate peak marker holds for <see cref="PeakHoldSeconds"/> then decays at its own
/// rate, and a clip latch sticks until explicitly reset. A mutable value type exactly as on the
/// Mac (`var left = AudioMeterChannelState()`) — keep instances in a plain field (not a property,
/// not an array element) so mutating calls apply in place rather than to a copy.
public struct AudioMeterChannelState
{
    public const float FloorDb = -60f;
    public const float CeilingDb = 0f;
    public const float LevelDecayDbPerSecond = 24f;
    public const float PeakDecayDbPerSecond = 18f;
    public const double PeakHoldSeconds = 1.5;

    private float _levelDb;
    private double _levelTime;
    private float _peakDb;
    private double _peakHoldUntil;

    public bool Clipped { get; private set; }

    public AudioMeterChannelState()
    {
        _levelDb = FloorDb;
        _peakDb = FloorDb;
    }

    public void Ingest(float peak, double time)
    {
        var current = Display(time);
        float incomingPeak = Decibels(peak);
        _levelDb = MathF.Max(incomingPeak, current.LevelDb);
        _levelTime = time;

        if (incomingPeak >= current.PeakDb)
        {
            _peakDb = incomingPeak;
            _peakHoldUntil = time + PeakHoldSeconds;
        }
        else if (time > _peakHoldUntil)
        {
            _peakDb = current.PeakDb;
            _peakHoldUntil = time;
        }
        Clipped = Clipped || peak >= 1f;
    }

    public readonly AudioMeterChannelDisplay Display(double time)
    {
        float levelElapsed = (float)Math.Max(0.0, time - _levelTime);
        float peakElapsed = (float)Math.Max(0.0, time - _peakHoldUntil);
        return new AudioMeterChannelDisplay(
            MathF.Max(FloorDb, _levelDb - levelElapsed * LevelDecayDbPerSecond),
            MathF.Max(FloorDb, _peakDb - peakElapsed * PeakDecayDbPerSecond),
            Clipped);
    }

    public void ResetClipping() => Clipped = false;

    public static float Decibels(float amplitude) => amplitude > 0f ? MathF.Max(FloorDb, 20f * MathF.Log10(amplitude)) : FloorDb;
}

/// Verbatim port of `Audio/AudioMeter.swift`'s `AudioMeterHub` — owns one <see cref="AudioMeterChannelState"/>
/// per stereo channel. The Mac's is `@MainActor`; this isn't thread-affine by construction, but
/// AudioMeterView (its only caller) only ever touches it from the UI thread, so the effective
/// contract is the same.
public sealed class AudioMeterHub
{
    private AudioMeterChannelState _left = new();
    private AudioMeterChannelState _right = new();

    public void Ingest(AudioMeterAnalysis analysis, double time)
    {
        _left.Ingest(analysis.LeftPeak, time);
        _right.Ingest(analysis.RightPeak, time);
    }

    public StereoAudioMeterDisplay Display(double time) => new(_left.Display(time), _right.Display(time));

    public void ResetClipping()
    {
        _left.ResetClipping();
        _right.ResetClipping();
    }

    public void Reset()
    {
        _left = new AudioMeterChannelState();
        _right = new AudioMeterChannelState();
    }
}
